storage/block.go

```diff
@@ -124,27 +122,26 @@ func (b *dbBlock) close() {
 		return
 	}
 
-	// Otherwise, we need to close the stream.
-	if stream := b.stream; stream != nil {
-		b.ctx.RegisterCloser(func() {
-			if bytesPool := b.opts.GetBytesPool(); bytesPool != nil {
-				segment := stream.Segment()
-				stream.Reset(m3db.Segment{})
-				bytesPool.Put(segment.Head)
-				bytesPool.Put(segment.Tail)
-			}
-			stream.Close()
-		})
-	}
+	// Otherwise, we need to return bytes to the bytes pool.
+	segment := b.segment
+	bytesPool := b.opts.GetBytesPool()
+	b.ctx.RegisterCloser(func() {
+		// NB(xichen): we do not return segment tail here because the tails
+		// are precomputed in the encoding options and therefore potentially
+		// reused across different encoders.
```

Comment:
Oh wow, this is a very subtle/complex one.  Good comment.  Wondering if we should perhaps encapsulate this somehow, e.g. add a method to `m3db.Segment` that is called `Release` so you can `segment.Release(bytesPool)` and it will nil out the `Head` and `Tail` and only return the `Head` to the `bytesPool`?

Also makes me wonder if we will release the bytes for reading data just a little bit too early as we close every fetch context at the end of the call inside the TChannel service, however after that function finishes the underlying TChannel implementation still needs to read those objects and encode them on the wire.  I'll have a talk with @prashantv and @martin-mao about this.